### DISCUSSION TITLE
Update percentage completion or any metadata for a task

### DISCUSSION THREAD

**Discussion by ChaitanyaTessell** [Netflix/conductor#2354] [Q&A] [answered]
I am using the conductor version v3.0.6.

I wanted to update the percentage complete for a task and some other data (in the output). If I call conductor's update task api from my worker with output as percentage value during the execution of the task, the conductor actually puts the task again in the queue to be polled. I just wanted to updated some output in the update task call and nothing else, the worker should just update the task and continue its execution.
Do we have any such functionality in conductor?

**Comment by kishorebanala** [comment, marked as answer]
@ChaitanyaTessell Task logs is a better place to do this. This is append only, compared to over-writing task output.

Updating the task releases the worker lease, and goes through another cycle of poll and execute on a different worker. Renewing the task lease on same worker and continuing the execution is not supported today. Should you be interested, feel free to open a discussion with your proposal.

**Reply by ChaitanyaTessell** [reply]
Task logs are more of an unparseable dump. I would not get structured data out of it.
I am more looking towards having variables (task and workflow level) which can be updated and shared anytime. (There is a support for variables at workflow level but I suppose, that can only be updated by executing some task).
For now, I have a usecase of having long running tasks where I wanted to show the end user about the progress in terms of percentage completion (and might have future usecase to update some more variables during execution).

**Reply by apanicker-nflx** [reply]
>I am more looking towards having variables (task and workflow level) which can be updated and shared anytime.

Updating task data out of band to facilitate progress tracking does not seem to be a compelling enough use-case to expose this to race conditions considering that this does not impact the actual execution and hence workflow execution in anyway.

>For now, I have a usecase of having long running tasks where I wanted to show the end user about the progress in terms of percentage completion 

Task execution logs precisely serve this purpose while being non-intrusive to the task execution. I understand that the unstructured format makes it difficult to parse, but a quick fix could look something like the optional string serialization for TaskSummary and WorkflowSummary using [SummaryUtil](https://github.com/Netflix/conductor/blob/main/common/src/main/java/com/netflix/conductor/common/utils/SummaryUtil.java).

**Reply by naik91** [reply]
Agreed on the part that non contexual stuff can be and should be put in the logs.
I have one more usecase for this. As I mentioned that I have long running tasks, where I will do multiple steps inside a single task, I want to store some data at the certain checkpoints inside a task, so that in case of retries I should be able to resume by accessing the same data again. Is that something that is possible or any way in which we can implement it?

**Reply by kishorebanala** [reply]
@naik91 In the long running task case, Is the task using `callBackAfterSeconds` to update and poll later? For eg., if the worker thinks it'll need to execute later, it can update Conductor with `IN_PROGRESS` state and some duration of `callBackAfterSeconds`. The task will then be available for poll after this duration, where another worker (say worker-2) will continue from where initial worker (say worker-1) left off. In this case, you can put any additional data in task output while updating and the subsequent execution can use it, if available.

However, if the worker is long running, and does not update Conductor with `callBackAfterSeconds`, thereby not giving up the lease, I think it's best to store any task specific state elsewhere than in Conductor.

**Reply by naik91** [reply]
Thanks for your response.
The worker itself is long running and I will need that data in case of cleanup workflow (failure workflow scenarios) and possibly in retries for the failed tasks. So, storing the context of workflow execution in the workflow itself makes sense for me.
Let me know your thoughts on this.

**Reply by kishorebanala** [reply]
@naik91 State updates outside of task lifecycle could get trickier. For eg., if the worker dies before the state update reaches Conductor, Conductor wouldn't be able to guarantee correctness of this state.

> As I mentioned that I have long running tasks, where I will do multiple steps inside a single task..

Would it be possible to split each of these steps into a separate task, consider using DO WHILE?

**Reply by naik91** [reply]
Got it.
Do while will not work as the steps are not the same and I want to keep them in a single task as it is logically a single task.
I think I got your point on data inconsistency as it would be something like a 2 phase commit (step execution + task data save), so saved data in task might not be always present.
Will think on the solve where it can be stored at the execution end.
Thanks for your responses.
